### DISCUSSION TITLE
Route modal - once route modal loaded , previous content getting wiped/replaced

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [remix-run/remix#7618] [Q&A] [answered]
> **NOTE:  Using Remix v1 routing system**

**Goal**
Trying to create a route base modal.
Currently using Remix v1 routes file system routing

**Background context**
1. A user is on `/dci/metro/visualizer/$device_name` route
2. User clicks device which adds a `/detail ` to the existing URL
   - before clicking device:  `/dci/metro/visualizer/$device_name`
   - after clicking device (which triggers modal route): `/dci/metro/visualizer/$device_name/detail`

```
├── dci
│  ├── metro
│  │  └── visualizer
│  │     └── ($device)
│  │        ├── detail.tsx (modal route)
│  │        └── index.tsx. (devices visualization)
```

**Challenge / Observations:**
Once the `/detai`l route is active AKA  we are now on this route  `/dci/metro/visualizer/$device_name/detail`,
The modal appears for brief half-second
Previous content gets wiped completely
Modal doesnt show

**Question:**
How do I get my route modal to appear directly over the previous content I had?
- Before modal route is active, the route is: `/dci/metro/visualizer/$device_name`
- After modal route is active, the route is: `/dci/metro/visualizer/$device_name/detail`

https://github.com/remix-run/remix/assets/6743796/cd84e154-48fe-427e-b05a-601d7966accc

**Comment by cliffordfajardo** [comment]
Ran `npx remix routes` and got this:


Ran `npx remix routes` and got this:

```tsx
<Routes>
  <Route file="root.tsx">
    <Route path="dci/metro/visualizer/:device?/detail" index file="routes/dci/metro/visualizer/($device)/detail/index.tsx" />
    <Route path="dci/metro/visualizer/:device?" index file="routes/dci/metro/visualizer/($device)/index.tsx" />
  </Route>
</Routes>
```

Looks like `/detail` route is not nested as I thought it was according to the generated route config :thinking: 

I was expecting the generated output to look like this:
```tsx
<Routes>
  <Route file="root.tsx">
    <Route path="dci/metro/visualizer/:device?" file="routes/dci/metro/visualizer/($device)/detail/index.tsx">
      <Route path="dci/metro/visualizer/:device?/detail" index file="routes/dci/metro/visualizer/($device)/index.tsx" />
    </Route>
    
  </Route>
</Routes>
```

This explains why my Outlet isn't working as expected 🤔

**Comment by sergiodxa** [comment]
```
├── dci
│  ├── metro
│  │  └── visualizer
│  │     └── ($device)
│  │        ├── detail.tsx (modal route)
│  │        └── index.tsx. (devices visualization)
```

Here detail and index are at the same level.

You need to either open the modal without using nested routes (remove detail) or if you want to make it linkable make detail a nested route of the visualization.

**Reply by cliffordfajardo** [reply]
One of my goals is for the modal (`/detail`( to be linkable and also have the modal have its own `loader` and `actions`
Is there a way to get the modal to show up at this route?


```
https://localhost:15037/dci/metro/visualizer/$device/detail (get modal to appear here, but show visualization still)
```

Currently the visualization is rendered at this route:
```
Filesystem:      https://localhost:15037/dci/metro/visualizer/$device/ 
Browser URL:  https://localhost:15037/dci/metro/visualizer/lor1-bbr02/ 
```

![CleanShot 2023-10-09 at 10 40 32](https://github.com/remix-run/remix/assets/6743796/96b1ade8-646d-4892-bb41-95a8c2013d98)


Trying to get the modal to show up as its own route (so it can have its own loader/actions)
```
Filesystem:      https://localhost:15037/dci/metro/visualizer/$device/details (want modal to show here, but keep visualization content)
Browser URL:  https://localhost:15037/dci/metro/visualizer/lor1-bbr02/details
```

![CleanShot 2023-10-09 at 10 41 02](https://github.com/remix-run/remix/assets/6743796/019b20db-f19a-4ac6-965d-4b49f0a6a07d)

For some reasons I thought `/detail` would be a nested route because I its after the $device param in the URL:

**What I thought would get generated from React Router**
```tsx
<Routes>
  <Route file="root.tsx">
    <Route path="dci/metro/visualizer/:device?" file="routes/dci/metro/visualizer/($device)/detail/index.tsx">
      <Route path="dci/metro/visualizer/:device?/detail" index file="routes/dci/metro/visualizer/($device)/index.tsx" />
    </Route>
    
  </Route>
</Routes>
```

**What is actually generated:**

```tsx
<!-- Current file structure generates this, according to `npx remix routes` -->
<Routes>
  <Route file="root.tsx">
    <Route path="dci/metro/visualizer/:device?/detail" index file="routes/dci/metro/visualizer/($device)/detail/index.tsx" />
    <Route path="dci/metro/visualizer/:device?" index file="routes/dci/metro/visualizer/($device)/index.tsx" />
    <Route path="*" file="routes/$.tsx" />
  </Route>
</Routes>

```

**Reply by cliffordfajardo** [reply]
I tried to make `/detail` (the route file containing the modal) a nested route of `/visualization`:

**routes folder:**
```
├── dci
│  ├── metro/
│  │  └── visualizer/
│  │     ├── detail/
│  │     │  └── index.tsx
│  │     └── index.tsx
```
```


 but the generated code according to `npx remix routes` is generating this:

```tsx
<Routes>
  <Route file="root.tsx">
  <Route path="dci/metro/visualizer" index file="routes/dci/metro/visualizer/index.tsx" />
  <Route path="dci/metro/visualizer/detail" index file="routes/dci/metro/visualizer/detail/index.tsx" />
</Routes>
```

Was expecting it to output something like this:
```
<Routes>
  <Route file="root.tsx">
    <Route path="dci/metro/visualizer"  file="routes/dci/metro/visualizer.tsx">
      <Route path="detail" file="routes/dci/metro/visualizer/detail.tsx" />
    </Route>
  </Route>
</Routes>
```

Going to play around with not using `index` as the file name, I think thats affecting the route generation

**Reply by cliffordfajardo** [reply]
UPDATE: 
Ok so I was playing around with different remix file/folder naming conventions and I think I've found my original issue. 
**I had an incomplete mental model for how the route generation worked**
Haven't played with remix v2 routes yet, but probably similar.



✅ Generates nested routes in the way I was originally looking for

<table>
<tr>
  <td><b>Folder structure</b></td> 
  <td> <b>Generated routes (`npx remix routes`)</b></td>
</tr>
<tr>
<td>

```
routes
├── dci
│  ├── metro/
│  │  ├── visualizer/
│  │  │  └── detail.tsx
│  │  └── visualizer.tsx
```
</td>

<td>

```tsx
<Routes>
  <Route file="root.tsx">
    <Route path="dci/metro/visualizer" file="routes/dci/metro/visualizer.tsx">
      <Route path="detail" file="routes/dci/metro/visualizer/detail.tsx" />
    </Route>
  </Route>
</Routes>
```
</td>
</tr>
</table>



❌ Doesnt generates nested routes as I originally thought (using `index.tsx`)
<table>
<tr>
  <td><b>Folder structure</b></td> 
  <td> <b>Generated routes (`npx remix routes`)</b></td>
</tr>
<tr>
<td>

```
routes
├── dci
│  ├── metro/
│  │  └── visualizer/
│  │     ├── detail/
│  │     │  └── index.tsx
│  │     └── index.tsx
```
</td>

<td>

```tsx
<Routes>
  <Route file="root.tsx">
  <Route path="dci/metro/visualizer" index file="routes/dci/metro/visualizer/index.tsx" />
  <Route path="dci/metro/visualizer/detail" index file="routes/dci/metro/visualizer/detail/index.tsx" />
</Routes>
```
</td>
</tr>
</table>

**Reply by kiliman** [reply]
Yes, in `v1` routing, _index.tsx_ generates an `index` route. In `v2` routing, they changed it to `_index.tsx`

Index routes can't have child routes. That's why there is no nesting.

**Reply by cliffordfajardo** [reply, marked as answer]
Updating the thread with working solution of the original question

**✅ Working code (doesnt make use of `index.tsx)**
<table>
<tr>
  <td><b>Folder structure</b></td> 
  <td> <b>Verify Generated routes (`npx remix routes`)</b></td>
</tr>
<tr>
<td>

```
├── dci
│  ├── index.tsx
│  ├── metro
│  │  └── visualizer
│  │     ├── ($device)
│  │     │  └── detail.tsx
│  │     └── ($device).tsx
```
</td>

<td>

```tsx
<Routes>
  <Route file="root.tsx">
    <Route path="dci/metro/visualizer/:device?" file="routes/dci/metro/visualizer/($device).tsx">
      <Route path="detail" file="routes/dci/metro/visualizer/($device)/detail.tsx" />
    </Route>
  </Route>
</Routes>
```
</td>
</tr>
</table>

![CleanShot 2023-10-09 at 12 26 17](https://github.com/remix-run/remix/assets/6743796/f01bd5ee-7c21-4a1f-97e9-c528ac0f4c9e)

<br /><br />



**❌ Previously incorrect code**
I was incorrectly using `index.tsx` thinking I would still be able to generated nested routes by simply nesting folders.
See @kiliman comment above for clairifaction:
> Yes, in v1 routing, index.tsx generates an index route. In v2 routing, they changed it to _index.tsx
> Index routes can't have child routes. That's why there is no nesting.

<table>
<tr>
  <td><b>Folder structure</b></td> 
  <td> <b>Generated routes (`npx remix routes`)</b></td>
</tr>
<tr>
<td>

```
routes
├── dci
│  ├── metro/
│  │  └── visualizer/
│  │     ├── detail/
│  │     │  └── index.tsx
│  │     └── index.tsx
```
</td>

<td>

```tsx
<Routes>
  <Route file="root.tsx">
  <Route path="dci/metro/visualizer" index file="routes/dci/metro/visualizer/index.tsx" />
  <Route path="dci/metro/visualizer/detail" index file="routes/dci/metro/visualizer/detail/index.tsx" />
</Routes>
```
</td>
</tr>
</table>
